### PR TITLE
proper session restoration

### PR DESCRIPTION
**Pull request by zwhitchcox**
Continuation of #1734.

>Currently, if you open a session after vim is already opened (not with `vim -S ./path/to/session.vim`), the old `nvim-tree` buffers are not restored. This ensures that `nvim-tree` is reloaded whenever a session is loaded.

I have tested this pretty thoroughly and made sure that it functions properly when opening from a previous session with or without an `NvimTree` buffer. It also remembers which buffer you were on correctly and focuses the correct file.

It also functions properly if you delete all buffers (e.g. with `%bd!`) in your current session and load a new session.

I personally use [`auto-session`](https://github.com/nvim-tree/nvim-tree.lua/pull/1734), which creates a new session, listening to global directory changes, also saving your previous session. So, by setting

```
  sync_root_with_cwd = true,
  actions = {
    change_dir = {
      global = true
    }
  },
```

You can switch around in `nvim-tree`, and it automatically restores your sessions as you go, which is, frankly, pretty cool.

I only mention that (other than the fact that it's cool) to note that there may be other session managers with corner cases I haven't noticed, but it's working really well with my set up.

The only limitation is that if you have multiple tabs with NvimTree buffers, it does not restore the tabs that were not in focus at the time you made the session.

Still, it's a lot of progress.

cc @kyazdani42 @alex-courtis (because you were pinged before on the previous PR)

### CODE REVIEW

**Comment by alex-courtis** [comment]
I like the approach of using `SessionLoadPost` however I think I am doing something wrong:

When running a minimal configuration `nvim -nu /tmp/nvt-min.lua -S` `SessionLoadPost` never fires. Is that being fired before we invoke nvim-tree setup?

**Comment by alex-courtis** [comment]
> The only limitation is that if you have multiple tabs with NvimTree buffers, it does not restore the tabs that were not in focus at the time you made the session.

That's not a regression. It would be great if you could make that work...

File: lua/nvim-tree/lib.lua [hunk 1]
```diff
@@ -113,12 +113,17 @@ end
 
 function M.open(cwd)
   M.set_target_win()
+  print("open x")
   if not core.get_explorer() or cwd then
+  print("no explorer")
     core.init(cwd or vim.loop.cwd())
   end
   if should_hijack_current_buf() then
+    print("should hijack")
```
Comment (alex-courtis) [inline lua/nvim-tree/lib.lua:122]: You can use logs e.g. `log.line("dev", "SessionLoadPost")` for easier debugging.

Set
```lua
	  log = {
		  enable = true,
		  truncate = true,
		  types = {
			  dev = true,
		  },
	  },
```

and tail `~/.cache/nvim/nvim-tree.log`

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Looks like this is very much a work in progress... I'll revisit once we are in a stable state.

**Comment by zwhitchcox** [comment]
>Looks like this is very much a work in progress... I'll revisit once we are in a stable state.


@alex-courtis Yes, but your feedback helps immensely, thank you!

**Comment by alex-courtis** [comment]
> > Looks like this is very much a work in progress... I'll revisit once we are in a stable state.
> 
> @alex-courtis Yes, but your feedback helps immensely, thank you!

Added [wiki page](https://github.com/nvim-tree/nvim-tree.lua/wiki/Development )

Any additions would be gratefully appreciated.

**Comment by zwhitchcox** [comment]
Ok, @alex-courtis this is ready for another review whenever you're up for it. It now restores sessions per tab as well as fixing your other recommendations. Sorry for the delay...just had our first child!

**Comment by alex-courtis** [comment]
It looks like we get a `SessionLoadPost` for each window (not buffer) opened wen `:so mysession.vim`:

```
[2022-11-29 12:39:29] [dev] nvim-tree.setup
[2022-11-29 12:39:29] [dev] nvim-tree.on_enter
[2022-11-29 12:39:33] [dev] SessionLoadPost {
  buf = 8,
  event = "SessionLoadPost",
  file = "/home/alex/.local/share/nvim/vundle/nvim-tree.lua.dev/NvimTree_1",
  group = 14,
  id = 6,
  match = "/home/alex/.local/share/nvim/vundle/nvim-tree.lua.dev/NvimTree_1"
}
[2022-11-29 12:39:33] [dev] reattach_session
[2022-11-29 12:39:33] [dev] SessionLoadPost {
  buf = 7,
  event = "SessionLoadPost",
  file = "/home/alex/.local/share/nvim/vundle/nvim-tree.lua.dev/LICENSE",
  group = 14,
  id = 6,
  match = "/home/alex/.local/share/nvim/vundle/nvim-tree.lua.dev/LICENSE"
}
[2022-11-29 12:39:33] [dev] reattach_session
[2022-11-29 12:39:35] [profile] START git toplevel /home/alex/.local/share/nvim/vundle/nvim-tree.lua.dev
[2022-11-29 12:39:35] [profile] END   git toplevel /home/alex/.local/share/nvim/vundle/nvim-tree.lua.dev  2ms
[2022-11-29 12:39:35] [profile] START git untracked /home/alex/.local/share/nvim/vundle/nvim-tree.lua.dev
[2022-11-29 12:39:35] [profile] END   git untracked /home/alex/.local/share/nvim/vundle/nvim-tree.lua.dev  2ms
[2022-11-29 12:39:35] [profile] START git job /home/alex/.local/share/nvim/vundle/nvim-tree.lua.dev nil
[2022-11-29 12:39:35] [profile] START git job /home/alex/.local/share/nvim/vundle/nvim-tree.lua.dev nil
[2022-11-29 12:39:35] [profile] END   git job /home/alex/.local/share/nvim/vundle/nvim-tree.lua.dev nil  4ms
[2022-11-29 12:39:35] [profile] START draw
[2022-11-29 12:39:35] [profile] END   draw  1ms
[2022-11-29 12:39:35] [profile] END   git job /home/alex/.local/share/nvim/vundle/nvim-tree.lua.dev nil  11ms
[2022-11-29 12:39:35] [profile] START draw
[2022-11-29 12:39:35] [profile] END   draw  0ms
```

git is debounced and draw is cheap, so that's OK. debouncing or attempting to enforce one session load is not worth it.

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Congrats on the family!

Tested OK: `:so mysession.vim`

Not working: `nvim -S mysession.vim`

nvim-tree is not catching the `SessionLoadPost` events in the latter case, as the session is loaded before nvim-tree is started. That is unavoidable.

It looks like we will need to retain the `find_existing_windows` hacks to load into the existing window. That area of functionality is problematic and will be replaced soon #1669

Please just do your best...

File: lua/nvim-tree/utils.lua [hunk 2]
```diff
@@ -38,6 +38,10 @@ function M.path_split(path)
   return path:gmatch("[^" .. path_separator .. "]+" .. path_separator .. "?")
 end
 
+function M.get_tabnr(name)
+  return tonumber(string.sub(name, string.find(name, "%d+")))
```
Comment (alex-courtis) [inline lua/nvim-tree/utils.lua:42]: This blows up when string.find returns nil.

**Comment by alex-courtis** [comment]
Closing due to lack of activity. Please reopen if you wish to continue.

**Comment by martin-braun** [comment]
This PR is sad. I'm a frequent user of `:mks!` and `source Session.vim`, but I constantly have to deal with an empty buffer in place of nvim-tree. Not so annoying that I would invest hours into fixing this as it's quickly fixed by tapping my show/hide keymap for nvim-tree twice to effectively reload nvim-tree, but so much so that is starts to annoy day by day just a tiny little bit.

**Comment by alex-courtis** [comment]
This is a very tricky one, with the variety of session mechanisms.

This is all we have for now... https://github.com/nvim-tree/nvim-tree.lua/wiki/Recipes#workaround-when-using-rmagattiauto-session

**Comment by kqvanity** [comment]
@alex-courtis  The current workaround only restore the nvim-tree pane not empty without its state. any updates on that?

**Comment by alex-courtis** [comment]
> @alex-courtis The current workaround only restore the nvim-tree pane not empty without its state. any updates on that?

Honestly, no. This is a very complex situation, with the session not properly persisting all the information needed.

There is work underway to properly support "plugin windows", see `:help winfixbuf`. I imagine that information could be persisted in the session, which would make it easier to restore the session. Could you perhaps do a little research into the current state of nvim 0.10 and 0.11 to see if there is any work done/underway?

What's the state of other plugin RE sessions? Which ones work, which ones don't?

**Comment by kqvanity** [comment]
> Could you perhaps do a little research into the current state of nvim 0.10 and 0.11 to see if there is any work done/underway?

It's probably gonna be API-related changes, so it might go over my head.

> What's the state of other plugin RE sessions? 

I've initially used persisted.nvim, and it doesn't even bother restoring the nvim-tree pane.

**Comment by alex-courtis** [comment]
> I've initially used persisted.nvim, and it doesn't even bother restoring the nvim-tree pane.

Honestly, that sounds like a good outcome and that it's intentional on the part of persisted.

Other session managers behave very poorly, restoring the tree window in a dead state.

**Pull request closed without merging**